resources/analysisTools/copyNumberEstimationWorkflow/phasing.sh

```diff
@@ -0,0 +1,89 @@
+#!/bin/bash/
+
+# Copyright (c) 2017 The ACEseq workflow developers.
+# Distributed under the MIT License (license terms are at https://www.github.com/eilslabs/ACEseqWorkflow/LICENSE.txt).
+
+#DEFINE FILE NAMES
+UNPHASED="${FILE_UNPHASED_PRE}${CHR_NAME}.${FILE_VCF_SUF}"
+UNPHASED_TWOSAMPLES="${FILE_UNPHASED_PRE}${CHR_NAME}_2samples.${FILE_VCF_SUF}"
+PHASED_TWOSAMPLES="${FILE_PHASED_GENOTYPE}${CHR_NAME}_2samples"
+tmpPhased="${FILENAME_PHASED_GENOTYPES}_tmp"
+tmpHaploblocks="${FILENAME_HAPLOBLOCK_GROUPS}_tmp"
+
+if [[ ${isNoControlWorkflow} == false ]]
+then
+
+    source ${TOOL_ANALYZE_BAM_HEADER}
+    getRefGenomeAndChrPrefixFromHeader ${FILE_CONTROL_BAM} # Sets CHR_PREFIX and REFERENCE_GENOME
+
+    CHR_NR=${CHR_PREFIX}${CHR_NAME:?CHR_NAME is not set}
+
+         ${SAMTOOLS_BINARY} mpileup ${CNV_MPILEUP_OPTS} -u \
+         	    -f "${REFERENCE_GENOME}" \
+         	    -r ${CHR_NR} \
+         	    "${FILE_CONTROL_BAM}" \
+         	    | \
+         	    ${BCFTOOLS_BINARY} view ${BCFTOOLS_OPTS} - \
+         	    > "${UNPHASED}"
+         
+         if [[ "$?" != 0 ]]
+         then
+         	echo "Non zero exit status for mpileup in phasing.sh" >> /dev/stderr
+         	exit 2
+         fi
+fi
+
+echo -n > "${UNPHASED_TWOSAMPLES}"
+echo -n > "${tmpPhased}"
+echo -n > "${tmpHaploblocks}"
+
+${PYTHON_BINARY} "${TOOL_BEAGLE_CREATE_FAKE_SAMPLES}" \
+    --in_file "${UNPHASED}" \
+    --out_file "${UNPHASED_TWOSAMPLES}"
+
+if [[ "$?" != 0 ]]
```

Comment:
Nice! Fail early with good message! Could be improved with

```bash
echo "Non zero exit status ($?) while creating 2nd sample in vcf-file in phasing.sh"
```

Not sure, but you may have to rescue the exit code in a variable, though:

```bash
${PYTHON_BINARY} "${TOOL_BEAGLE_CREATE_FAKE_SAMPLES}" \
￼     --in_file "${UNPHASED}" \
￼     --out_file "${UNPHASED_TWOSAMPLES}" \
     && sample2Ec="$?"
￼
if [[ "$sampel2Ec" != 0 ]]; then
  echo "Non zero exit status (=$sample2Ec) while creating 2nd sample in vcf-file in phasing.sh" >> /dev/stderr
￼ 	exit 2
fi 
```

Reply:
The code a bit repetitive. You could do:

```bash
dieWith() {
  local msg="${1:?No error message}"
  local ec="${2:-$?}"
  echo "$msg: exit code '$ec'" >> /dev/stderr
  exit 2  # or `exit "$ec"`
}

command1 || dieWith "my message"

command2 || dieWith "other message"
```

Reply:
That's a cool idea.
Just as a note: `command1 || dieWith "my message"` does not work but `command1 || dieWith "my message" $?` does. Maybe bash cannot access `$?` inside the function?

Reply:
Below is similar code I have in the alignment workflow. This worked there without any problems.

```bash
throw () {
  local lastCommandsExitCode=$?
  local exitCode="${1-$UNSPECIFIED_ERROR_CODE}"
  local msg="${2-$UNSPECIFIED_ERROR_MSG}"
  if [[ $lastCommandsExitCode -ne 0 ]]; then
    msg="$msg (last exit code: $lastCommandsExitCode)"
  fi
  errout "$exitCode" "$msg"
  printStackTrace
  exitIfNonInteractive "$exitCode" || return $?
}
```

Maybe it is because the "$?" reference is not in the first line? Could you try
```bash
dieWith() {
  local lastExitCode="$?"
  local msg="${1:?No error message}"
  local ec="${2:-$lastExitCode}"
  echo "$msg: exit code '$ec'" >> /dev/stderr
  exit 2  # or `exit "$ec"`
}
```

I tried it on the command line and it worked.

```bash
dieWith() {   local lastExitCode="$?";   local msg="${1:?No error message}";   local ec="${2:-$lastExitCode}";   echo "$msg: exit code '$ec'" >> /dev/stderr; echo exit 2; }

$ true || dieWith "message"
$ false || dieWith "message"
message: exit code '1'
exit 2
```

Reply:
> Maybe it is because the "$?" reference is not in the first line? Could you try

Yes that's it! Otherwise `$?` references the exit code of the assignment ` local msg="${1:?No error message}"`

Reply:
What is the advantage of using
```bash
dieWith() {
  local lastExitCode="$?"
  local msg="${1:?No error message}"
  local ec="${2:-$lastExitCode}"
  echo "$msg: exit code '$ec'" >> /dev/stderr
  exit 2  # or `exit "$ec"`
}
```
compared to
```bash
dieWith() {
  local ec="${2:-$?}"
  local msg="${1:?No error message}"
  echo "$msg: exit code '$ec'" >> /dev/stderr
  exit 2  # or `exit "$ec"`
}
```
Is it simply readability?